### PR TITLE
[LSP] Support for sequential formatting with multiple clients

### LINKED ISSUE #14371: lsp: support multiple formatting providers
`formatting()` (and `range_formatting()` for that matter) currently requests `document/formatting` from all capable clients at once and applies the results (text edit array) once they arrive. There are two main problems with that approach:

1. `document/formatting` responses are applied to the buffer without checking whether the buffer has changed since the request has been made. But the buffer could have been changed by the `document/formatting` response of a different server, or even by user input.
2. the order in which responses are applied is undefined.

## Use Case

Personally I have the efm general purpose language server installed which provides `document_formatting`. Often times I’d like to use *both* efm *and* another language server to provide formatting. Some common use cases are:

1. JSON: formatting form json language server + formatting from efm (with fixjson)
2. HTML: formatting from html language server + formatting from efm (rustywind, a class sorter for tailwindcss)
3. Blade: formatting from html language server + formatting from efm (blade-formatter and rustywind)

## Possible Solution

I could make a PR to add the following function, which applies formatters in a chain, ie applies formatters one at a time. I assume changes to the code base will be purely additive. I’m happy to discuss the details here :)

`vim.lsp.buf.formatting_chain_sync({options}, {timeout_ms}, {order})`

Requests formatting from all capable servers, ie those that provide document_formatting or document_range_formatting. The formatting requests are send one at a time such that the formatting request to the next server is only send after the response of the previous server has been processed.

### Parameters:

**{options}** (optional, table) like in `vim.lsp.buf.formatting_sync({options}, {timeout_ms})`

**{timeout_ms}** (optional, number) like in `vim.lsp.buf.formatting_sync({options}, {timeout_ms})`

**{order}** (optional, table) list of client names used to determine the order in which formatting requests are send

### Example:

Suppose we have 4 clients attached:

1. efm: `document_formatting`
2. html: `document_range_formatting`
3. foo: `document_formatting`
4. bar: no formatting

`vim.lsp.buf.formatting_chain_sync(nil, nil, { 'html', 'efm' })`
_request order:_ foo → html → efm

`vim.lsp.buf.formatting_chain_sync(nil, nil, { 'efm' })`
_possible request order:_ foo → html → efm
_possible request order:_ html → foo → efm

`vim.lsp.buf.formatting_chain_sync(nil, nil, nil)`
html, foo and efm _with undefined request order_

### Related
https://github.com/neovim/neovim/issues/12865
https://github.com/neovim/neovim/pull/13233

### PR DESCRIPTION
**Pull request by karimknaebel**
Fixes https://github.com/neovim/neovim/issues/14371

### CODE REVIEW

File: runtime/lua/vim/lsp/buf.lua [hunk 1]
```diff
@@ -144,6 +145,48 @@ function M.formatting_sync(options, timeout_ms)
...
+        break
+      end
+    end
+  end
+
+  -- loop through the clients and make synchronous formatting requests
+  for _, client in ipairs(clients) do
+    if client.resolved_capabilities.document_formatting then
+      local result = client.request_sync("textDocument/formatting", vim.lsp.util.make_formatting_params(options), timeout_ms)
+      print(client.name)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/buf.lua:182]: I assume this was for debugging and should be removed?

File: runtime/lua/vim/lsp/buf.lua [hunk 2]
```diff
@@ -144,6 +145,48 @@ function M.formatting_sync(options, timeout_ms)
...
 
+--- Formats the current buffer by sequentially requesting formatting from attached clients.
+---
+--- Useful when multiple clients with formatting capability are attached.
+---
+--- Since it's synchronous, can be used for running on save, to make sure buffer is formatted
+--- prior to being saved. {timeout_ms} is passed on to the |vim.lsp.client| `request_sync` method.
+--- Example:
+--- <pre>
+--- vim.api.nvim_command[[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting_chain_sync()]]
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/buf.lua:156]: This name seems out of date

**Comment by mjlbach** [comment]
Other than my comments this seems fine to me. @tjdevries @mfussenegger any other thoughts? Mostly this is geared toward the tsserver + efm crowd

File: runtime/lua/vim/lsp.lua [hunk 3]
```diff
@@ -890,6 +890,42 @@ function lsp.start_client(config)
...
+    local request_result = nil
+    local function _sync_handler(err, _, result)
+      request_result = { error = err, result = result }
+    end
+
+    local success, request_id = client.request(method, params, _sync_handler,
+      bufnr)
+    if not success then return nil end
+
+    local wait_result, reason = vim.wait(timeout_ms or 100, function()
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:918]: Is the timeout_ms high enough for prettier and the like? I don't use this functionality, so I wasn't positive

**Comment by kabouzeid** [comment]
I've tested this for two weeks now, with various language servers, works reliable for me.

As a side note: In my local setup I fall back to Document Range Formatting on the whole buffer when Document Formatting is not available (this is the case for `html`, `css` and `json` language servers from VSCode). This will come in a separate PR since we need to decide whether it should be opt in or not. You can see it in the original issue: https://github.com/neovim/neovim/issues/14371#issuecomment-820706468.

**Comment by mfussenegger** [comment]
I wonder if this new `M.formatting_seq_sync` shouldn't replace the existing `formatting_sync`?
Is there a case where using the existing `formatting_sync` variant would be prefered for multiple clients? If there's a single client, both of them behave the same anyways, unless I'm missing something.

**Comment by mjlbach** [comment]
I don't think there's an instance where you would use `formatting_sync` over `formatting_seq_sync` with multiple clients

**Comment by kabouzeid** [comment]
I agree, right now there isn't really a case to use the old `formatting_sync`. To quote myself from what I posted in the original issue, these are the problems with the current API and what would an improved API could look like:

---

## Current API

`formatting()`

Requests formatting from all clients, with current state of buffer.  
*Problem*: all formatters use the same buffer state → they can interfere with each other

`formatting_sync()`

Requests formatting from all clients, with current state of buffer. Waits for all clients to respond, but then only picks the first result.  
*Problem*: not the same behaviour as the async call → confusing  
*Problem*: first result could be empty, but there might be a non-empty result that we could apply

`range_formatting()`

Requests range formatting from all clients, with current state of buffer.  
*Problem*: all formatters use the same buffer state → they can interfere with each other


## What should be the API (changes in bold)

`formatting()`

Requests formatting from all clients, with current state of buffer. **Only applies the first non-empty result.**

`formatting_sync()`

Requests formatting from all clients, with current state of buffer. Waits for all clients to respond, **only applies the first non-empty result.**

`range_formatting()`

Requests range formatting from all clients, with current state of buffer. **Only applies the first non-empty result.**

`range_formatting_sync()` **new**

Requests range formatting from all clients, with current state of buffer. Only applies the first non-empty result.

`formatting_seq()` **new**

Requests formatting from all clients, one at a time, always with the updated state of buffer, so that all formatters can be applied at once without interfering with each other.

`formatting_seq_sync()` **new**

Requests formatting from all clients, one at a time, always with the updated state of buffer, so that all formatters can be applied at once without interfering with each other.


## Notes

It would not make sense to also have `range_formatting_chain` as this can never work reliably.

I would argue it makes sense to always fall back to range_formatting on the whole buffer if formatting is not available. This way we support VSCode’s html, css, json lang servers and possibly others.

**Comment by mjlbach** [comment]
In what case would you only want to apply the first result? Isn't that somewhat random depending on the order the clients are initialized?

**Comment by kabouzeid** [comment]
It's random in the sense which client first gets a response.

The current implementation of `formatting_sync` only applies the first result, and the new specification would be a "bug fix" for the current behavior. But you are right, I also don't see a use case for it when we also have `formatting_seq_sync`.
The only upside of this would be that it works analogous to `formatting_range_sync`, which can never be applied in sequence.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp.lua:918]: According to @fsouza's blog post (https://blog.fsouza.dev/prettierd-neovim-format-on-save/) prettier takes about 350ms. So I guess changing the default `timeout_ms` to either 500ms or 1000ms should be plenty.

**Comment by mjlbach** [comment]
The other option might be grouping formatting based on number of clients rather than blocking behavior. So `formatting` and `formatting_chain`. If you call formatting, and there are multiple clients attach, we prompt you to select which client you would like to use. If you call formatting_chain, we replicate the behavior of this PR.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:918]: I would test, but I imagine that 500ms is fine, as I guarantee there are people who are going to be using this with prettier.

**Comment by mfussenegger** [comment]
Another option would be to add a client predicate to both `formatting` and `formatting_sync`. If not nil it would only use the clients matching the predicate.

If users want to use multiple clients for formatting, they'd have to call the functions multiple times. Ordering would be implicit with `formatting_sync` due to the order of the function calls.

**Comment by mjlbach** [comment]
What is the utility of an async variant with multiple clients though, as their buffer states will be branched at the same point, so their edits will be incompatible.

**Comment by mjlbach** [comment]
Ah, yes, then I think I agree with the split of the 4 functions, I worry a bit about not blocking during sequential formatting operations as the buffer may become desynchronized though due to user input.

**Comment by kabouzeid** [comment]
@mjlbach your worries about the async variant are true, but this problem exists already with the current async `formatting`. Can probably be fixed in both cases by checking the change tracking version before applying the edits?

**Comment by mjlbach** [comment]
Maybe, I would say for now just implement formatting_seq_sync() matching your current implementation, and change formatting()/formatting_sync() as we discussed and we can have a separate discussion about how to make the async variant of formatting sync robust. (That is, if @mfussenegger agrees with the plan). I think this is actually fairly close to your current PR with some name changes and the additional multi-server checks.

**Comment by mfussenegger** [comment]
> Also, now that I think about it again, in what situation would you use `formatting_sync` over `formatting_seq_sync`?

That was my initial question :)
I'd adjust `formatting_sync` to behave like `formatting_seq_sync` instead of adding a new function

**Comment by kabouzeid** [comment]
I see :) I can't think of a case where you only want to run a single formatter when multiple are available, am I missing something? For misbehaving clients that you never want as formatters, you can always set the document formatting capability to false.

**Comment by mjlbach** [comment]
I just think the API becomes much more clear is the sync part implies that the formatting is blocking, and the seq part implies the formatting is part of a sequence (for multiple formatters). I imagine some people probably prefer EFM or tsserver for example, and may want to run different formatters on different files in parts of their project.

**Comment by mjlbach** [comment]
It's not useless, because if you only have 1 formatter, then it would not prompt you. If you have multiple formatters, you can set up your own logic for invoking the correct autocommand or use formatting_seq_sync.

**Comment by kabouzeid** [comment]
I'm still not quite sure that I'm convinced, but in the end I'll also be happy with your proposal :)

Do you agree though that `formatting` should always ask which formatter to use when there are more than one? i.e. the users choice will not be remembered.

**Comment by mjlbach** [comment]
Yeah, I don't think we should remember user's choice, if they want to have that, they can implement their own caching logic by overriding the function (should be like a 3 liner)

**Comment by kabouzeid** [comment]
Done. Seems to be working fine.

File: runtime/lua/vim/lsp/buf.lua [hunk 4]
```diff
@@ -111,6 +111,39 @@ function M.completion(context)
...
+    end
+    local user_choice = vim.fn.confirm(
+      "Multiple formatters available, which one would you like to use?",
+      table.concat(choices, "\n"),
+      0,
+      "Question"
+    )
+    if user_choice == 0 then return nil end
+    return clients[user_choice]
+  elseif #clients < 1 then
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/buf.lua:140]: Can't we do this check earlier? , like around line 120?

**Comment by mjlbach** [comment]
Nice! Looks great :) I'll go through and leave a more detailed review later, but I didn't see any red flags? @mfussenegger what do you think?

File: runtime/lua/vim/lsp/buf.lua [hunk 5]
```diff
@@ -111,6 +111,39 @@ function M.completion(context)
...
+  -- better UX when choices are always in the same order (between restarts)
+  table.sort(clients, function (a, b) return a.name < b.name end)
+
+  if #clients > 1 then
+    local choices = {}
+    for k,v in ipairs(clients) do
+      table.insert(choices, string.format("%d %s", k, v.name))
+    end
+    local user_choice = vim.fn.confirm(
+      "Multiple formatters available, which one would you like to use?",
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/buf.lua:133]: This sounds a bit too wordy to me (nitpick). Maybe, "Select a language server for formatting:"

Comment (kabouzeid) [reply runtime/lua/vim/lsp/buf.lua:133]: Agreed

File: runtime/lua/vim/lsp/buf.lua (hunk 4 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp/buf.lua:140]: The earliest would be line 124, which would be immediately before this `if` branch. Having it here also emphasizes that `#clients == 1` holds for the `else` case.

Comment (mjlbach) [reply runtime/lua/vim/lsp/buf.lua:140]: That's fine, I think it's a little more readable to return early if there are no clients as it matches some of the rest of the API but it's totally your call

**Comment by mjlbach** [comment]
I approved your test run, if either mfusseneger or @tjdevries sign off I'll merge (trying to do the Rule of Twos for reviews that change any pre-existing functionality)

**Comment by mjlbach** [comment]
Oh and please squash your commits when ready :)

**Comment by kabouzeid** [comment]
Will do

**Comment by kabouzeid** [comment]
Tests were passing. Probably should've waited until the very end before squashing and force pushing 🤔

File: runtime/lua/vim/lsp.lua [hunk 6]
```diff
@@ -890,6 +890,42 @@ function lsp.start_client(config)
...
+  --@param method (string) LSP method name.
+  --@param params (table) LSP request params.
+  --@param timeout_ms (number, optional, default=100) Maximum time in
+  ---milliseconds to wait for a result.
+  --@param bufnr (number) Buffer handle (0 for current).
+  --@returns { err, result }, where `err` and `result` come from the |lsp-handler|.
+  ---On timeout, cancel or error, returns `(nil, err)` where `err` is a
+  ---string describing the failure reason. If the request was unsuccessful
+  ---returns `nil`.
+  --@see |vim.lsp.buf_request_sync()|
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp.lua:907]: The docs in `vim.lsp.client` should probably be extended so that this method shows up as well.

That there are 3 error scenarios also bothers me a bit, but I can see that it could be useful to distinguish between all of them, so I guess it's okay that way.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua:918]: :+1: for increasing the default to 500ms

File: runtime/lua/vim/lsp/buf.lua [hunk 7]
```diff
@@ -134,14 +170,58 @@ end
...
-  local result = vim.lsp.buf_request_sync(0, "textDocument/formatting", params, timeout_ms)
-  if not result or vim.tbl_isempty(result) then return end
-  local _, formatting_result = next(result)
-  result = formatting_result.result
-  if not result then return end
-  vim.lsp.util.apply_text_edits(result)
+  local result = client.request_sync("textDocument/formatting", params, timeout_ms)
+  if result and result.result then
+    util.apply_text_edits(result.result)
+  end
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/buf.lua:182]: Should we also give the users some feedback if there is an error. Maybe using `vim.notify` with the error message?

File: runtime/lua/vim/lsp.lua [hunk 8]
```diff
@@ -890,6 +890,42 @@ function lsp.start_client(config)
...
+  --- Sends a request to the server and synchronously waits for the response.
+  ---
+  --- This is a wrapper around {client.request}
+  ---
+  --@param method (string) LSP method name.
+  --@param params (table) LSP request params.
+  --@param timeout_ms (number, optional, default=100) Maximum time in
+  ---milliseconds to wait for a result.
+  --@param bufnr (number) Buffer handle (0 for current).
+  --@returns { err, result }, where `err` and `result` come from the |lsp-handler|.
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp.lua:903]:
```suggestion
  --@returns { err, result }, where `err` and `result` come from the |lsp-handler|.
```

By reading the docs I thought this would be a list where the first item is the error and the second item is the result. But reading the code I figured it is a dictionary with `err` and `result` as keys. Maybe that could be clarified in the docs?

**Review by mfussenegger — COMMENTED** [review]
Left some more minor comments, overall looks good to me.

File: runtime/lua/vim/lsp.lua (hunk 6 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp.lua:907]: Yes it's a bit weird. I did this to stay as close as possible to `buf_request_sync` which also has those 3 error scenarios; when none of the requests are successful it returns `nil` (and until recently it would return `{}`).

I interpret this as: returns `nil` when something went wrong, and returns an optional second value `err` with details about the error.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp.lua:918]: Will test today with prettier and other slower formatters to see if 500 or 750 are sane defaults.

File: runtime/lua/vim/lsp/buf.lua (hunk 7 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp/buf.lua:182]: Makes sense to me. Probably a good idea for any API function that has a `timeout_ms` parameter.

File: runtime/lua/vim/lsp.lua (hunk 8 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp.lua:903]: Indeed. It's the same format as the `request_result` from `buf_request_all` and `buf_request_sync`, where the structure is not documented at all.

Something like `{ ["error"] = error, ["result"] = result }` or `{ error = error, result = result }` is verbose but should make things clear. What do you think?

**Comment by kabouzeid** [comment]
`range_formatting` now also works like the updated `formatting`

File: runtime/lua/vim/lsp/buf.lua (hunk 7 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp/buf.lua]: Done

File: runtime/lua/vim/lsp.lua (hunk 6 above)
Comment (kabouzeid) [reply runtime/lua/vim/lsp.lua]: Added doc in `vim.lsp.client`

**Comment by kabouzeid** [comment]
All done now. I've increased the default timeout from _100ms_ to _1000ms_. The slowest formatters I could find were prettier (~350ms) and blade-formatter (~750ms).

**Review by mfussenegger — APPROVED** [review]

**Comment by mjlbach** [comment]
I'm going to merge, I think this is a strict improvement over status quo, and we can get feedback on the new interface leading up to 0.5, and possibly do another round of revisions based on @mfussenegger's suggestions if needed

**Comment by mjlbach** [comment]
Ah crap, I wanted these squashed. My mistake...
